### DISCUSSION TITLE
Undefined error when using HTTPS with next-auth

### DISCUSSION THREAD

**Discussion by skrenes** [nextauthjs/next-auth#2318] [Help]
**Describe the bug**
I'm running Next.js on my own infrastructure with HTTPS, which as far as I known requires setting up the NodeJS server manually. When I do this, however, I get an `undefined` error with next-auth when attempting email authentication.

**Steps to reproduce**
1. Use the vanilla setup by following the [example instructions](https://next-auth.js.org/getting-started/example).
2. create a `server.js` file to enable HTTPS although the problem can be reproduced using HTTP as well if you do not have certificates available. Refer to **Additional context** for example code for both.
3. Start server with command `node server.js`
4. Sign in with email authentication
5. Check email and click "Sign in" button
6. Error is displayed

**Expected behavior**
I should be authenticated. Instead I get a black page with the heading "Error" and my domain as a subheading. With the next-auth debug option turned on, the console reports (email in console is changed to email@domain.com):
```
[next-auth][debug][typeorm_get_user_by_email] email@domain.com
[next-auth][debug][typeorm_create_verification_request] email@domain.com
[next-auth][debug][typeorm_get_verification_request] undefined undefined
```

**Screenshots or error logs**
N/A

**Additional context**
```node
// server.js with HTTP
const { createServer } = require('http');
const next = require('next');

const hostname = process.env.HOSTNAME || 'localhost';
const port = parseInt(process.env.PORT, 10) || 3000;
const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(() => {
  createServer((req, res) => {
    const parsedUrl = new URL(req.url, `http://${hostname}:${port}`);
    handle(req, res, parsedUrl);
  }).listen(port, (err) => {
    if (err) throw err;
    console.log(`> Ready on http://${hostname}:${port}`);
  });
});
```

```node
// server.js with HTTPS
const { createServer } = require('https');
const next = require('next');
const fs = require('fs');

const hostname = process.env.HOSTNAME || 'localhost';
const port = parseInt(process.env.PORT, 10) || 3000;
const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

const options = {
  key: fs.readFileSync('/cert/server.key'),
  cert: fs.readFileSync('/cert/server.crt'),
};

app.prepare().then(() => {
  createServer(options, (req, res) => {
    const parsedUrl = new URL(req.url, `https://${hostname}:${port}`);
    handle(req, res, parsedUrl);
  }).listen(port, (err) => {
    if (err) throw err;
    console.log(`> Ready on https://${hostname}:${port}`);
  });
});
```

**Feedback**
*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [ ] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [x] Could not find relevant documentation
* [ ] Found the example project helpful
* [x] Did not find the example project helpful

**Comment by iaincollins** [comment]
Have you defined `NEXTAUTH_URL` as an environment variable (ie `process.env.NEXTAUTH_URL`)?

With a setup like the above you should be able to do it at runtime in `server.js`, but it needs to be defined as an environment variable so that it's exported everywhere on the app (on all routes, and both client and server side).

**Comment by skrenes** [comment]
Yes that environment variable is set. Just to sanity check, I disabled it and not only did I get a console warning about the missing variable but the signin callback broke (since I have a FQDN instead of `localhost`) and no email was sent.

**Comment by skrenes** [comment]
Any ideas? Doesn't proper authentication in a production environment necessitate HTTPS? How is this done when not hosting with Vercel? I sure hope it's not to run a server in front of the server or some other redundant solution like a reverse proxy.

**Comment by nstolmaker** [comment]
@skrenes Did you figure out a solution?

**Comment by skrenes** [comment]
@nstolmaker Unfortunately I had no success solving this problem. So I ripped out the custom server (`server.js`) and put a reverse-proxy with HTTPS in front of it and that worked. I think this issue is worthy of solving. It's just beyond my ability as I'm starting off with full-stack web development.

**Comment by marcelobaez** [comment]
@nstolmaker Can you provide me your nginx config ? I'm having issues with next auth behind a reverse proxy

**Comment by skrenes** [comment]
I'm disappointed this has moved to a discussion, instead of being taken seriously.
